opm/simulators/wells/StandardWell_impl.hpp

```diff
@@ -1520,24 +1532,37 @@ namespace Opm
 
         // calculate the phase rates based on the primary variables
         // for producers, this is not a problem, while not sure for injectors here
-        if (well_type_ == PRODUCER) {
+        if (this->isProducer()) {
             const double g_total = primary_variables_[WQTotal];
             for (int p = 0; p < number_of_phases_; ++p) {
                 well_state.wellRates()[index_of_well_ * number_of_phases_ + p] = g_total * F[p];
             }
         } else { // injectors
-            // TODO: using comp_frac_ here is very dangerous, since we do not update it based on the injection phase
-            // Either we use distr (might conflict with RESV related) or we update comp_frac_ based on the injection phase
             for (int p = 0; p < number_of_phases_; ++p) {
-                const double comp_frac = comp_frac_[p];
-                well_state.wellRates()[index_of_well_ * number_of_phases_ + p] = comp_frac * primary_variables_[WQTotal];
+                well_state.wellRates()[index_of_well_ * number_of_phases_ + p] = 0.0;
+            }
+            switch (this->wellEcl().injectorType()) {
+            case Well::InjectorType::WATER:
+                well_state.wellRates()[index_of_well_ * number_of_phases_ + Water] = primary_variables_[WQTotal];
+                break;
+            case Well::InjectorType::GAS:
+                well_state.wellRates()[index_of_well_ * number_of_phases_ + Gas] = primary_variables_[WQTotal];
+                break;
```

Comment:
I tried to run this PR locally on [`SPE1CASE2_OILGAS.DATA`](https://github.com/OPM/opm-tests/blob/b7e972f35e82f9a79d82be531b04cd38496fe664/spe1/SPE1CASE2_OILGAS.DATA) and got a valgrind error here.  Assigning to
```
wellRates()[... + Gas]
```
looks unsafe to me.  I would expect there to be something resembling `pu.phase_pos` here.

Reply:
Thanks a lot, that was indeed a bug!